### DISCUSSION TITLE
关于开发者模式下的日志打印

### DISCUSSION THREAD

**Discussion by LonelySteve** [the1812/Bilibili-Evolved#3105] [功能建议 / Ideas]
其实如果非要在目前的代码中实现的话，也不是做不到，只是写起来感觉特别丑：

```javascript
import { getGeneralSettings } from '@/core/settings'

getGeneralSettings().devMode && console.error('error message')
```

而且这种打日志的方式比较随意，缺乏一些作为日志最基本的要素（范围、时间、事件级别等）

简单考虑一下下面这种使用形式：

```javascript
import { createLogger } from '@/core/logging'

const logger = createLogger('scope name, eg. component name')

logger.debug('debug message')  // 只在开发者模式下显示
logger.info('info message')
logger.warn('warn message')
```

不一定非得自己创建 logger，组件和插件的 `entry` 函数中可以拿到自动生成 logger 对象

```javascript
entry: ({ logger }) => {
  logger.debug('debug message')
}
```

`createLogger` 中封装一些时间，事件级别等信息的格式化，最终调用 console.log 等接口打印出来，甚至可以考虑在开发者模式下，将日志存在配置或者 localStorage 中以便翻阅.

看大家有什么想法或建议，我觉得类似上面这种就基本足够使用了.

**Comment by the1812** [comment]
- 时间和级别, `console` 都支持, 时间要在 DevTools 里打开
- 范围可能不好确定, 通过 `entry` 注入的话组件关联的其他文件不好拿到, 而且我希望能像 `console` / `logError` 一样随处可访问
- `unsafeWindow.console` 和 `window.console` 是不同的对象, 或许可以直接向 `window.console` 上修改一些函数, 比如 `console.debug` 改成仅在 devMode 下打印

**Reply by LonelySteve** [reply]
- 时间和级别，`console` 是都支持，但是我希望本体打印的日志能够和普通的`console`打印的区分开来，封装到 `logger` 里便可以统一实现，B 站自己其实可能就会打印一坨日志，但是对咱们分析错误几乎没有帮助，还容易误导小白以为是 Bilibili-Evolved 的报错，发到 issue 里
- 什么情况下需要让其他文件拿到通过 `entry` 注入的 `logger` 对象呢？我觉得最多让这个组件/插件文件自己能拿到就可以了，非要跨文件共享的话，考虑下我写的 [useShare](https://github.com/the1812/Bilibili-Evolved/blob/bbfbefad29cc86f762200c586249f013bac2539e/registry/lib/components/video/player/common/speed/utils.ts#L5-L28) ？
- 我能够理解第三点的想法，如果能实现的话，改动量很小，而且对开发者透明，可惜难以实现统一的范围管理

**Reply by the1812** [reply]
> B 站自己其实可能就会打印一坨日志

如果只是区分脚本和 b 站的日志, 感觉也可以通过改写 `console.log` 来插入一段固定的前缀

> 什么情况下需要让其他文件拿到通过 entry 注入的 logger 对象呢？

稍微复杂一些的组件都会分多个文件的, 比如自定义顶栏, 分出来的文件里面才是主要逻辑, 但并不直接在 `entry` 的作用域里, 必须要用到像你说的 `useShare` 之类的共享机制.
所以如果需要精确到组件的范围的话, 好像还是可以在 `console` 上搞点文章(

```ts
import { scopedConsole } from '@/core/logger'

const console = scopedConsole('componentName')

console.log('a = 1')
// => [Bilibili Evolved] [componentName] a = 1

```

**Reply by LonelySteve** [reply]
其实 `console`/`logger` 在不考虑 hack `window.console` 的情况下，最大区别只是叫法不同啦，前面的风格可以更贴近原生 API，使用起来心智负担更小，为了方便我就都换成 `console` 表述了.

我稍微改了下你的代码：

```diff
- import { scopedConsole } from '@/core/logger'
+ import { scopedConsole } from '@/core/console'

- const console = scopedConsole('componentName')
+ export const console = scopedConsole('componentName', { internal: true })

console.log('a = 1')
// => [Bilibili Evolved][internal(with color)][componentName] a = 1
```

要用的话，可以单独作为一个文件导出 `console` 对象，在组件/插件范围内导入使用，也不一定非要用 `useShare` 实现，用 `useShare` 实现最大的优点是可以实现在全局范围内共享相同的 `console` 对象，不过一般用不着跨组件/插件共享吧？这会破坏范围定义，最好只在本组件/插件内共享，而且安全获取 `console` 的时序也是个问题...

> 稍微复杂一些的组件都会分多个文件的, 比如自定义顶栏

其实我原先的想法是从 `entry` 里拿到的 `console`/`logger` 逐层向下传递，不过要挂载 Vue 组件的话，这个传递可能会有点困难/丑陋，那就干脆不把 `console` 注入 `entry` 了， 直接从 `./console.ts` 导入组件/插件共享的 `console`.

可能唯一一点不爽的是，这样设计需要重复写一遍 componentName（

**Reply by the1812** [reply]
`{ internal: true }` 是什么作用?

**Reply by LonelySteve** [reply]
> `{ internal: true }` 是什么作用?

请忽略这个，，😂

我本意是想让内置的组件/插件打印的日志能和其他组件/插件有点区别，比如加上额外的前缀或者换上更暗淡的颜色，但是昨天我想的这个方式其实，也可以被普通组件用上，那就失去意义了（

再思考一段时间吧，如果没有想到更好的设计，我就先把倍速的PR发布了，之后设计好了再补上.

**Comment by the1812** [comment]
`console.debug` 在 DevTools 里属于 Verbose 级别, 默认是不开启的, Hook 的时候是不是换成调用 `console.log` 方便使用些

**Reply by LonelySteve** [reply]
## 对于开发者来说

虽然默认情况下，`console.debug` 打印出来的日志不显示（只是隐藏起来了，不是不打印），但是可以通过切换 Level 筛选不同级别的日志，这个比起敲 Filter 筛选总归是要舒服一些的.

反过来说，Filter 可以设定为 bilibili-evolved 的日志前缀，这样就排除了 B 站以及浏览器自己的一些无关日志，配合 Level 更方便筛选查看.

## 对于用户来说

日志模块选择 `console.log` 或对应原生级别实现都无所谓，本身让用户按下 F12 打开开发者工具，切到 Console 选项卡，就有一定的操作成本（不过上 B 站的都是年轻人，感觉都会弄233）.

我的想法是，在发起 issue 时，比起上传日志截图，上传文本日志会更好，github 大图片加载会有点慢，而且可能存在日志没有截全的问题.

本体可以在日志模块中集成记录保存日志的功能，这个和你要不要用 `console.log` 实现 `console.debug` 无关，在调用真正的打印接口之前，如果开启了开发者模式，就把日志保存到某个地方，将来在界面上提供一个按钮以供下载，以后发起 issue 时，就可以提示用户上传近期的日志以供分析.

推荐将日志保存到 `SessionStorage` 里，这样既可以涵盖绝大部分 BUG 出现的情景，也可以减少不必要日志的存储.

## 总的来说

我觉得还是按原生 API 一对一实现即可，不要统一使用 `console.log`.

**Reply by LonelySteve** [reply]
保存日志这个功能可以之后再考虑，可能不是很必要，devtools 自己可以保存日志文件的，而且这样保存下来的日志是最全的。

我这样设计只是降低了用户的操作成本...

**Comment by timongh** [comment]
我同意将日志保存起来，并提供复制或下载功能。再复杂一点，甚至可以根据时间、组件、大小进行筛选。这样用户提 issue 方便，我们看日志也方便。

那么为了保存日志，日志 api 应该仅接受文本，不应该依赖原生 console 的对象打印功能。

**Reply by the1812** [reply]
那不至于, 兼容 JSON 的数据都可以保存

**Comment by timongh** [comment]
### 一

我认为 logger 一定要有随意添加 scope 的能力。并且可以在一个 logger 上叠加  scope 成为一个新 logger。

这样可以方便组件作者对组件代码进行更细粒度的分类，使其更容易找到自己关注的部分。

### 二

应该支持生成随机 ID 作为 scope。

由于事件的存在，我们的日志可能会非常混乱。

例如：某组件注册了一个点击事件，如果用户点击了很多次，那么每次点击所输出的日志就会混杂在一起，我们根本无法区分哪些日志来自于哪一次点击。（补充：如果只有同步代码，由于 js 是单线程，那么单次点击的日志还是会放在一起。但如果有异步代码，比如发出了网络请求，那么代码的执行顺序就说不准了）

有一个固定的 ID 作为区分，我们才能确认某个日志属于哪个调用栈。

在具体实现上，我有两种方案：

1. 跟上述内容一样，那么日志看起来应该是这样：`[scope1|d83kv|scope3|8wv4o|d3j2p] a long message`。即让 ID 看起来像个普通  scope。
2. 不仅允许 ID 作为 scope，同时也允许普通 scope 绑定一个随机 ID。就像这样：`[scope1|scope2:6sok1|se813] a long message.` 这么做的好处就是更直观

当然，作为另一种选择，也可以将生成随机 ID 的工作交给组件开发者自己。日志 API 只负责添加 scope。

**Comment by timongh** [comment]
在打印日志时，我们常常有打印键值对的需求。因此我建议 api 可以像这样：

```ts
// 仅打印消息
log('abcd')
// 打印消息，同时打印键值对，下面两种二选一
log('the message', { k: v })
log({ msg: 'the message', k: v })
```

**Comment by timongh** [comment]
在打印日志时，我们常有拼接字符串的需求。如果只允许传入字符串，那么有可能造成性能浪费。

例如：我们在打印 debug 日志时拼接了字符串，而此时，在设置中却仅允许打印 warn 级别的日志。那么字符串就白拼接了。

我能想到两种解决方案：
1. 允许传入返回字符串的闭包，仅在打印时才从闭包中获取字符串。
2. 接收模板字符串和模板参数，仅在打印前完成拼接

如果采纳了前一条建议，即允许传入对象作为键值对参数。那么上述缺点也许能得到减轻或者消除。不过这只是我的猜测，我并不清楚在有了键值对参数后，是否还会有拼接字符串的需求。

**Comment by timongh** [comment]
日志打印仅按照 devMode 进行区分，我认为不够。或许应该在界面中添加打印级别的设置选项。

**Comment by the1812** [comment]
我统一回复下上面几条:
- scope 是支持的, 生成随机 ID 的 scope 我可以加上试试
- 拼接字符串是很快的, 得看用来拼接的参数, 会有为了产生一个只有日志用的数据而需要很重的计算的情况吗
- 级别和原生 console 对齐的, debug / log (info) / warn / error, 只有 debug 级别的才受 devMode 的开关影响

**Comment by timongh** [comment]
update: zone.js 的链接 url

我似乎把异步日志想得太简单了。。

我发现只是在日志中加个 ID，并不能解决异步日志的问题。因为在异步过程中会调用其他函数，而被调用的函数是无法获取这个异步 ID 的；除非在每个函数外面都包裹一层，并在调用时传一个上下文进去，这显然不太现实。

经过搜索，我发现有一个 [zone.js 库](https://github.com/angular/angular/tree/main/packages/zone.js) 可以帮助获取异步上下文。其实现方式大概是直接修改全局函数。

我再研究研究，看看整合到现有的 console 里面麻不麻烦。不是很麻烦我就搞一下。

**Reply by the1812** [reply]
没太明白, 有简单的例子看看吗

**Reply by timongh** [reply]
假设有如下代码：

```typescript
document.body.addEventListener('click', onClick)
document.body.addEventListener('keydown', onKeydown)

function onClick() {
    console.log('onClick')
    const val = something()
    common(val) // promise
}

function onKeydown() {
    console.log('onKeydown')
    const val = something()
    common(val) // promise
}

async function common(n: number) {
    console.error(`common 参数不能大于 256. n: ${n}`)
}
```

如果 `common` 是一个同步函数，那么当看到其错误日志的时候，如果我们想知道是哪个函数传入了不合法的参数，我们可以往日志的前面去找，看看这句日志跟在哪个后面就很容易找到了。如果前面的日志是 `onClick` 那肯定是 `onClick` 函数有问题；如果是 `onKeydown` 也是一个道理。

但 `common` 是异步的时候就不一样了，假如我们收到了这样的日志：
```plain
onClick
onKeydown
common 参数不能大于 256. n: 1000
```
我们无法判断这个出了错的 `common` 流程，是从 `onClick` 那边调用的，还是 `onKeydown` 那边调用的。于是不得不把两段代码都拿来看一遍。

我以前提的那个加一个随机 ID 作为 Scope 的建议，面对这种情况根本毫无作用。除非在 `common` 的参数里显式地加上一个参数，作为打印日志的 key。那总不能给每个需要打印日志的异步函数都硬加一个参数吧。

这就是为什么需要有感知异步上下文的 API，这个 API 应该由运行环境给出，比如 nodejs API。从这个异步上下文中，我们就能知道函数的调用关系。至于这个关系以什么形式体现就要看 API 怎么实现了。

我知道的不多，其中可能有的东西没说对，不过意思是这么个意思。

**Reply by timongh** [reply]
看了看，有点麻烦的样子，不好弄……所以，再说吧

**Reply by the1812** [reply]
那为啥不看堆栈呢
```ts
(() => {
  const console = useScopedConsole('test')

  async function common() {
    Promise.resolve().then(() => console.error('test error'))
  }

  unsafeWindow.testA = () => {
    console.log('testA')
    common() // promise
  }

  unsafeWindow.testB = () => {
    console.log('testB')
    common() // promise
  }
})()
```
![image](https://user-images.githubusercontent.com/26504152/196195865-48f713d9-b100-4dbb-a7f9-3554f7e5eb5a.png)

**Reply by timongh** [reply]
确实没意识到还有这个东西。对这个项目应该是够用了，大概。

有一点确实是挺不好意思的，之前我对这方面了解不深，有点想当然，提议往代码里加了个随机 ID 的 scope。现在看来确实是没什么用。代码里应该没什么地方在用这个功能，我在考虑要不干脆把这个东西删了。

**Comment by FoundTheWOUT** [comment]
我觉得现在这个 log 设计有一个地方不合理，那就是现在所有的 console 会在 log 这个文件内执行，准确来说
https://github.com/the1812/Bilibili-Evolved/blob/7173f9ca936e9798c1b3bbe762285c447caa6cb9/src/core/utils/log.ts#L100-L122
是在这个函数里执行了，这会导致在 devtools 里丢失很关键的行信息，让定位相关代码变得相对比较繁琐。可能返回一个 console 的原生对象（且修改过的）比较好。这样做的影响是会使 `ScopedConsoleCallHook` 失效，但现在貌似并没有使用到这个钩子的地方，而且这个钩子是否需要是不是也可以讨论一下？

**Reply by the1812** [reply]
主要是因为 `console.log` 只有第一层调用栈, 想看完整的可以用 `trace` / `warn` / `error`
